pkg/cdi/cdi.go

```diff
@@ -0,0 +1,102 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2022 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cdi
+
+import (
+	"fmt"
+
+	"github.com/container-orchestrated-devices/container-device-interface/pkg/cdi"
+	cdiSpecs "github.com/container-orchestrated-devices/container-device-interface/specs-go"
+	"github.com/golang/glog"
+
+	"github.com/k8snetworkplumbingwg/sriov-network-device-plugin/pkg/types"
+)
+
+// CreateCDISpec creates CDI spec file with specified devices
+func CreateCDISpec(resourcePrefix string, rPool types.ResourcePool) error {
+	prefix := resourcePrefix
+	if prefixOverride := rPool.GetResourcePrefix(); prefixOverride != "" {
+		prefix = prefixOverride
+	}
+	cdiDevices := make([]cdiSpecs.Device, 0)
+	cdiSpec := cdiSpecs.Spec{
+		Version: cdiSpecs.CurrentVersion,
+		Kind:    prefix + "/" + rPool.GetPoolKind(),
+		Devices: cdiDevices,
+	}
+
+	for _, dev := range rPool.GetDevices() {
+		containerEdit := cdiSpecs.ContainerEdits{
+			DeviceNodes: make([]*cdiSpecs.DeviceNode, 0),
+			Env:         make([]string, 0),
+		}
+
+		envs, err := rPool.GetEnvs(prefix, []string{dev.GetID()})
+		if err != nil {
+			glog.Errorf("CreateCDISpec(): Unable to create environment variables: %s", err)
+			return fmt.Errorf("unable to create environment variables: %s", err)
+		}
+		for name, value := range envs {
```

Comment:
also how will env variables will look in container ?

we may need to:
1. create CDI spec per allocate call
2. update CDI spec if exists in list and watch

Reply:
@adrianchiris I didn't understand your comment. Could you please describe what do you mean?

Reply:
These are general questions regarding this package and behaviour in general

1, you are creating a CDI spec in this package right ? how does a CDI spec file look like ?
2. when you allocate device to container through CDI what will be the env vars defined in container


1 and 2 when there are multiple devices per resource pool and per container allocation.

Reply:
1. CDI spec file example could be found here: https://github.com/container-orchestrated-devices/container-device-interface#full-blown-cdi-specification
2. With CDI I we pass exactly the same env vars as we do as a device plugin mode now.
3. annotation key and values are generated bu cdi library and I don't want to provice an example in code comments because it could be outdated once we bump cdi version. For a several devices, we'll have following annotation: `example.com/kind=example.com/kind=0000:00.00.1,example.com/kind=0000:00.00.2`

Reply:
@e0ne to keep compatibility with how currently device plugin exposes env variables, one option is to use current device plugin mechanism for env vars and just move device spec to CDI.

alternative that i can think of, is to generate a CDI device kind per allocation but thats hacky.

Reply:
environment variables are fixed in the latest version